### PR TITLE
Services register creates and returns a nested data structure

### PR DESCRIPTION
**Pull request by mboneil10**
Testing services_controller's register method

### CODE REVIEW

**Comment by ennukee** [comment]
~~It would be much easier on my end if you could have the register send an http request to the engine's receiving endpoint rather than putting everything in the response~~ (edit: on second look through this isn't what happens so it's fine, but one of the two do need to happen so the ~~router~~ engine gets data, preferably the request to the endpoint)

Just so I don't have to write code for both the initializer receiving data _and_ the engine update endpoint receiving data. It would be easier just to direct it all to the one endpoint.

(I also don't know how freely I can modify engine data because of the initializer's existence as an independent file in the `config/initializers` directory). It could potentially cause lag on the microservice's startup since it may wait for the initializer to finish (i.e. receive a response and parse it) if it has to parse a large response.

But I don't know how if it's considered poor practice to do an entirely new http request inside of a controller action (thoughts @dfaulken?)

**Comment by ennukee** [comment]
On the actual code, I'm surprised @dfaulken didn't mention this but shouldn't the `.require` syntax be used for the `url` param since based on the code I'm assuming the `url` is going to be mandatory parameter?

I _thiiink_ you may also need exception catching for the `Model.create(...)` for the cases we do get a non-unique model name (or using `.new` and testing it with `.valid?` would work too since `rescue` seems intensive)

**Comment by dfaulken** [comment]
@mboneil10 and I discussed on Slack how this test isn't accurate - this should be a case where there's a no-op.

I wasn't going to focus on much else until the code made that test pass, but any exceptions to the service register process can just halt app startup. I'm fine with that.

**Comment by ennukee** [comment]
When you say "halt app startup" are you referring to the microservice that's using this?

**Comment by dfaulken** [comment]
Yep. A service whose configuration doesn't pass muster with the router should need to be reconfigured before being able to register.

**Comment by mboneil10** [comment]
Forgot the nested data structure part. Fixing that now.

**Comment by ennukee** [comment]
On something I mentioned before, it seems like `.require` is more for requests done within the system, it gets much more finicky when you try to do it where it's an outside source sending requests (or that was my experience when setting up a test endpoint to play with).

**Comment by werebus** [comment]
@enragednuke, I don't think that's true.  Certainly, there's nothing "within the system" about a browser out there on the Internet, and those are the typical consumers of stand-alone Rails apps over HTTP.

As long as HTTP is the transport medium, there's no way to determine whether the client is a web browser or some other application and `.require`d parameters should work fine.  This is a key mechanism of identifying a valid request.

**Comment by ennukee** [comment]
@werebus Right, I wouldn't think it's any different, but I was saying that because when I was writing a dummy app to play with my initializer on the engine end, it didn't work using `params.require(...)` but it did with `params[...]`. Could've just been something silly missing (I think it said something about strong parameters? or a google search did), but that's just what I ran into.

**Comment by mboneil10** [comment]
@enragednuke, since params.require(...) could have failed on your dummy app for reasons unrelated to what we're doing here, I think we should wait. If we end up having to fix this later on, you have every right to say "I told you so."

File: app/controllers/services_controller.rb [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
-    @models = params[:models]
-    @models.each{ |m| Model.create(service: @service)}
-    render json: @service, 
-              except: %i(created_at updated_at id),
-              include: :models
+    service = Service.find_or_create_by url: params.require(:url)
+    params.require(:models).each do |model_data|
+      model = Model.find_by name: model_data.require(:name)
+      if model.present?
+        head :unprocessable_entity and return if model.service != service
```
Comment (mboneil10) [inline app/controllers/services_controller.rb:7]: @dfaulken, rubocop recommended the change made on line 7 (making two lines into one line). I'm not sure if this is what we want.

**Comment by dfaulken** [comment]
In general I only defy rubocop if I disagree with the rule as a whole - which I don't in this instance. If an `if` statement and its result can fit on a single line, I think it's generally true that those are simple enough ideas to be understood in a more compact format.

**Comment by ennukee** [comment]
> I think it's generally true that those are simple enough ideas to be understood in a more compact format.

Ternary operator: "Why won't you love me?!"

**Comment by dfaulken** [comment]
Ternary statements mandate expressing an `else`, which I don't favor doing inline in Ruby in any format.

**Comment by ennukee** [comment]
It was mostly a joke 😛

**Review by dfaulken — APPROVED** [review]
Looks great

**Pull request merged**